xcode/Subconscious/Shared/Parsers/Subtext.swift

```diff
@@ -45,6 +45,13 @@ struct Subtext: Hashable, Equatable, LosslessStringConvertible {
     var description: String {
         String(base)
     }
+    
+    /// Return a snapshotted String of the contents of `blocks`, joined by newlines.
+    func toString() -> String {
+        blocks
+            .map { b in b.span.toString() }
+            .joined(separator: "\n")
+    }
```

Comment:
No, because in the instance of `Subtext.truncate` we create a `Subtext` instance that uses the same `base` but only includes _some_ of the blocks. This avoids re-allocating the string but if we read directly from `base` we will get the original document not the filtered set of blocks.

We _could_ just create a new `base` String and point all the blocks at that though? I got the impression from the API we should avoid directly reading from `base` _because_ not every char in `base` has to correspond to a `block`.

Reply:
@bfollington `base` is a Substring rather than a string. This means multiple Subtext instances may share the same underlying string, but slice from different parts of it.

The Substring in `Subtext` should always correspond 1:1 with the parsed dom. We make use of this invariant in a number of ways.

We attempt to enforce this via the API. This is the reason that both `base` and `blocks` are `let` properties, and also the reason that `init(base:blocks:)` is a private constructor (it's an internal shortcut for the parser).

We want to make sure we uphold this invariant. I think we should get rid of `init(base:blocks:)` and force everything through the string/substring code path. The footgun is not worth any marginal perf gain we may have had through avoiding a re-parse.

In the case of `excerpt()`, I think that we should take the relevant blocks, stringify them, pass the resulting string to `Subtext(markup:)` and allow that to re-parse (it's cheap). This means that the excerpt doesn't need to hold on to the full document in memory, since its `base` points to the stringified excerpt.

Reply:
SGTM, I'll drop Subtext.toString after that and use the description instead.